bootstore/src/schemes/v0/peer.rs

```diff
@@ -0,0 +1,802 @@
+                handle_unique_id_counter: 0,
+                shutdown: false,
+                accepted_connections: BTreeMap::new(),
+                negotiating_connections: BTreeMap::new(),
+                connections: BTreeMap::new(),
+                rx,
+                init_responder: None,
+                rack_secret_responder: None,
+                log: log
+                    .new(o!("component" => "bootstore", "peer_id" => id_str)),
+                conn_rx,
+                conn_tx,
+            },
+            PeerHandle { tx },
+        )
+    }
+
+    /// Run the main loop of the peer
+    ///
+    /// This should be spawned into its own tokio task
+    pub async fn run(&mut self) {
+        // select among timer tick/received messages
+        let mut interval = interval(self.config.time_per_tick);
+        interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+        let listener = TcpListener::bind(&self.config.addr).await.unwrap();
+        while !self.shutdown {
+            tokio::select! {
+                res = listener.accept() => self.on_accept(res).await,
+                Some(request) = self.rx.recv() => {
+                    self.on_api_request(request).await;
+                }
+                Some(msg) = self.conn_rx.recv() => self.on_conn_msg(msg).await,
+                _ = interval.tick() => {
+                    if let Err(errors) = self.fsm.tick(Instant::now().into()) {
+                        for (_, err) in errors {
+                            self.handle_api_error(err).await;
+                        }
+                    }
+                    // Even with errors there may be messages that need sending
+                    self.deliver_envelopes().await;
+                }
+            }
+        }
+    }
+
+    // Handle an accepted connection
+    async fn on_accept(
+        &mut self,
+        res: Result<(TcpStream, SocketAddr), std::io::Error>,
+    ) {
+        match res {
+            Ok((sock, addr)) => {
+                let SocketAddr::V6(addr) = addr else {
+                    warn!(self.log, "Got connection from IPv4 address {addr}");
+                    return;
+                };
+                // TODO: Log if a peer with a lower address connects?
+                // Remove any existing connection
+                self.remove_accepted_connection(&addr).await;
+                info!(self.log, "Accepted connection from {addr}");
+                self.handle_unique_id_counter += 1;
+                let handle = spawn_server(
+                    self.handle_unique_id_counter,
+                    self.config.id.clone(),
+                    self.config.addr.clone(),
+                    addr.clone(),
+                    sock,
+                    self.conn_tx.clone(),
+                    &self.log,
+                )
+                .await;
+                self.accepted_connections.insert(addr, handle);
+            }
+            Err(err) => {}
```

Comment:
This seems suspicious to me, worth adding a comment explaining why it's okay?

Reply:
Actually, this is not ok. I'm going to log an error for lack of a better thing to do.

Reply:
Fixed in 851cc853f